test/hw_tests/hwtest_zoneset_switching.cpp

```diff
@@ -0,0 +1,104 @@
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+#include <ros/ros.h>
+#include <gtest/gtest.h>
+#include <future>
+
+#include <string>
+
+#include <std_msgs/Byte.h>
+#include <std_msgs/UInt8.h>
+
+#include "psen_scan_v2_standalone/util/expectations.h"
+
+namespace psen_scan_v2_test
+{
+static constexpr int32_t WAIT_FOR_MESSAGE_TIMEOUT_S{ 5 };
+static constexpr uint8_t SCANNER_POWER_BIT{ 4 };
+static constexpr uint8_t ZONE_SWITCH_BIT{ 8 };
+
+class ScanComparisonTests : public ::testing::Test
+{
+public:
+  void SetUp() override  // Omit using SetUpTestSuite() for googletest below v1.11.0, see
+                         // https://github.com/google/googletest/issues/247
+  {
+    ros::NodeHandle pnh{ "~" };
+    pub_relay_cmd_ = pnh.advertise<std_msgs::Byte>("/relay_cmd", 1);
+  }
+
+  void TearDown() override
+  {
+    sendRelayCmd(SCANNER_POWER_BIT);
+  }
+
+protected:
+  void sendRelayCmd(uint8_t cmd);
+
+protected:
+  ros::Publisher pub_relay_cmd_;
+};
+
+inline void ScanComparisonTests::sendRelayCmd(uint8_t cmd)
+{
+  std_msgs::Byte command;
+  command.data = cmd;
+  pub_relay_cmd_.publish(command);
+  ros::topic::waitForMessage<std_msgs::Byte>("/relay_cmd", ros::Duration(WAIT_FOR_MESSAGE_TIMEOUT_S, 0));
+}
+
+static constexpr std::chrono::seconds DEFAULT_TIMEOUT{ 3 };
+static constexpr std::chrono::milliseconds NOT_SET_TIMEOUT{ 5 };
+
+TEST_F(ScanComparisonTests, simpleCompare)
+{
+  ros::NodeHandle nh;
+  std::promise<void> received_active_zone_one_promise;
+  auto received_active_zone_one = received_active_zone_one_promise.get_future();
+  std::promise<void> received_active_zone_two_promise;
+  auto received_active_zone_two = received_active_zone_two_promise.get_future();
+
+  boost::function<void(const std_msgs::UInt8&)> callback = [&](const std_msgs::UInt8& msg) {
+    if (msg.data == 0 && received_active_zone_one.wait_for(NOT_SET_TIMEOUT) == std::future_status::timeout)
+    {
+      received_active_zone_one_promise.set_value();
+    }
+    else if (msg.data == 1 && received_active_zone_two.wait_for(NOT_SET_TIMEOUT) == std::future_status::timeout)
+    {
+      received_active_zone_two_promise.set_value();
+    }
+  };
+  auto zone_subscriber = nh.subscribe<std_msgs::UInt8>("/laser_1/active_zoneset", 10, callback);
+
+  sendRelayCmd(SCANNER_POWER_BIT);
+  EXPECT_FUTURE_IS_READY(received_active_zone_one, DEFAULT_TIMEOUT);
```

Comment:
I don't think it works like that. At least not deterministically. The relay cmd turns on the scanner (which it already is? Maybe we have to establish "rules" about which assumptions tests can make about the scanner status, if any.). Now you would have to wait until the scanner has booted and is started by the driver. Then you would have to wait for a first monitoring frame to be processed.

The `waitForMessage` call doesn't help you here. I am also curious about the callback implementation. I think you should stick to `async_barrier.h`.

Reply:
I only want to make sure, that the scanner switches to zoneset 0. I expect the scanner to be powered and don't turn him off in this test. I introduced a change, that hopefully makes this more clear to the reader. The barrier to my understanding would not block the scanner any more than the wait for future does in this case. If the scanner is disabled and just starts with this command due to an unexpected test precondition, then the test will fail but that is true for every hw test we have so far isn't it?